src/core/textures/Texture.js

```diff
@@ -452,51 +454,85 @@ export default class Texture extends EventEmitter
-        texture.textureCacheId = name;
+        BaseTexture.addToCache(texture.baseTexture, name);
+        Texture.addToCache(texture, name);
 
         // also add references by url if they are different.
         if (name !== imageUrl)
         {
-            BaseTextureCache[imageUrl] = baseTexture;
-            TextureCache[imageUrl] = texture;
+            BaseTexture.addToCache(texture.baseTexture, imageUrl);
+            Texture.addToCache(texture, imageUrl);
         }
 
         return texture;
     }
 
     /**
-     * Adds a texture to the global TextureCache. This cache is shared across the whole PIXI object.
+     * Adds a Texture to the global TextureCache. This cache is shared across the whole PIXI object.
      *
      * @static
      * @param {PIXI.Texture} texture - The Texture to add to the cache.
-     * @param {string} id - The id that the texture will be stored against.
+     * @param {string} id - The id that the Texture will be stored against.
      */
-    static addTextureToCache(texture, id)
+    static addToCache(texture, id)
     {
-        if (!texture.textureCacheId)
+        if (id)
         {
-            texture.textureCacheId = id;
+            if (!texture.textureCacheId)
+            {
+                texture.textureCacheId = id;
+            }
+
+            // @if DEBUG
+            /* eslint-disable no-console */
+            if (TextureCache[id])
+            {
+                console.warn(`Texture added to the cache with an id [${id}] that already had an entry`);
+            }
+            /* eslint-enable no-console */
+            // @endif
+
+            TextureCache[id] = texture;
         }
-        TextureCache[id] = texture;
     }
 
     /**
-     * Remove a texture from the global TextureCache.
+     * Remove a Texture from the global TextureCache.
      *
      * @static
-     * @param {string} id - The id of the texture to be removed
-     * @return {PIXI.Texture} The texture that was removed
+     * @param {string|PIXI.Texture} texture - id of a Texture to be removed, or a Texture instance itself
+     * @return {PIXI.Texture|null} The Texture that was removed
      */
-    static removeTextureFromCache(id)
+    static removeFromCache(texture)
     {
-        const texture = TextureCache[id];
+        if (typeof texture === 'string')
+        {
+            const textureFromCache = TextureCache[texture];
 
-        delete TextureCache[id];
-        delete BaseTextureCache[id];
+            if (textureFromCache)
+            {
+                textureFromCache.textureCacheId = null;
+                delete TextureCache[texture];
 
-        return texture;
+                return textureFromCache;
+            }
+        }
+        else
+        {
+            for (const prop in TextureCache)
+            {
+                if (TextureCache[prop] === texture)
+                {
+                    delete TextureCache[prop];
+                }
+            }
```

Comment:
Is this loop because a texture can be stored under multiple `textureCacheIds`? Could we just use that property to lookup the texture rather than looping through all textures?

If a texture can have multiple ids, just make `textureCacheId` an array and when it is added via `addToCache()` just add the new ID.